src/coreclr/vm/greenthreads.cpp

```diff
@@ -16,18 +16,114 @@ FCIMPL0(void, JIT_GreenThreadMoreStack)
 {
 }
 FCIMPLEND
+
+bool GreenThreadHelpersToSkip(TADDR code) { return false; }
+
+void InitGreenThreads() {}
+
+bool stackPointerLessThan(Thread* greenThread, TADDR sp1, TADDR sp2)
+{
+    return sp1 < sp2;
+}
+
 #else // FEATURE_GREENTHREADS
-struct GreenThreadData
+
+GVAL_IMPL(TADDR, g_lowForceStackUnwind1);
+GVAL_IMPL(TADDR, g_lowForceStackUnwind2);
+GVAL_IMPL(TADDR, g_highForceStackUnwind1);
+GVAL_IMPL(TADDR, g_highForceStackUnwind2);
+
+bool GreenThreadHelpersToSkip(TADDR code)
 {
-    StackRange osStackRange;
-    uint8_t* osStackCurrent;
-    uint8_t* greenThreadStackCurrent;
-    Frame* pFrameInGreenThread;
-    Frame* pFrameInOSThread;
-    GreenThreadStackList *pStackListCurrent;
-    bool greenThreadOnStack;
-    SuspendedGreenThread* suspendedGreenThread;
-};
+    if (code >= g_lowForceStackUnwind1 && code < g_highForceStackUnwind1)
+        return true;
+    if (code >= g_lowForceStackUnwind2 && code < g_highForceStackUnwind2)
+        return true;
+    return false;
+}
+
+bool stackPointerLessThan(Thread* thread, TADDR sp1, TADDR sp2)
+{
+    PTR_GreenThreadData greenThread = NULL;
+    if (thread != nullptr)
+        greenThread = thread->m_greenThreadData;
+
+    if ((greenThread == nullptr) || greenThread->pStackListCurrent == nullptr)
+    {
+        return sp1 < sp2;
+    }
+
+    PTR_GreenThreadStackList cursor = greenThread->pStackListCurrent;
+    while (true)
+    {
+        if (cursor->prev == nullptr)
+        {
+            break;
+        }
+        cursor = cursor->prev;
+    }
+    /*
+     * We divide the stack into blocks as follow:
+     *
+     * OS Stack            # n    <- Should have sp < osStackCurrent
+     * GreenThreadSegment  # n-1  <- right most (i.e. next most) element in the pStackList
+     * GreenThreadSegment  # ...
+     * GreenThreadSegment  # 1    <- left most (i.e. prev most) element in the pStackList
+     * OS Stack            # 0    <- osStackCurrent
+     *
+     */
+    int blockNumber = 1;
+    int b1 = 0;
+    int b2 = 0;
+    while (cursor != nullptr)
+    {
+        TADDR stackLimit = (TADDR)cursor->stackRange.stackLimit;
+        TADDR stackBase = (TADDR)cursor->stackRange.stackBase;
+        if (stackLimit < sp1 && sp1 <= stackBase)
+        {
+            b1 = blockNumber;
+        }
+        if (stackLimit < sp2 && sp2 <= stackBase)
+        {
+            b2 = blockNumber;
+        }
+        cursor = cursor->next;
+        blockNumber++;
+    }
+    TADDR osStackCurrent = (TADDR)greenThread->osStackCurrent;
+    if (b1 == 0 && sp1 < osStackCurrent)
+    {
+        b1 = blockNumber;
+    }
+    if (b2 == 0 && sp2 < osStackCurrent)
+    {
+        b2 = blockNumber;
+    }
+    if (b1 == b2)
+    {
+        return sp1 < sp2;
+    }
+    else
+    {
+        // The smaller the block number, the earlier it was created, the larger it is "logically"
```

Comment:
I think you could
- remove the pre-walk through `prev`
- start blockNumber at -1 and decrement it
- iterate through `prev` instead of `next`
- flip the direction on this `b2 < b1`

I bring it up because that's what I spent my time on reviewing this, but also it wasn't that hard either.